libqrexec/libqrexec-utils.h

```diff
@@ -98,4 +101,96 @@ static inline size_t max_data_chunk_size(int protocol_version) {
         return MAX_DATA_CHUNK_V3;
 }
 #define ARRAY_SIZE(s) (sizeof(s)/sizeof(s[0]))
+
+/* Replace all non-printable characters by '_' */
+void do_replace_chars(char *buf, int len);
+
+/* return codes for handle_remote_data and handle_input */
+#define REMOTE_EXITED -2
+#define REMOTE_ERROR  -1
+#define REMOTE_EOF     0
+#define REMOTE_OK      1
+
+/*
+ * Handle data from vchan. Sends MSG_DATA_STDIN and MSG_DATA_STDOUT to
+ * specified FD (unless it's -1), and MSG_DATA_STDERR to our stderr.
+ *
+ * Return codes:
+ *   REMOTE_EXITED - remote process terminated, do not send more data to it
+ *     ("status" will be set)
+ *   REMOTE_ERROR - vchan error occured
+ *   REMOTE_EOF - EOF received, do not access this FD again
+ *   REMOTE_OK - maybe some data processed, call again when buffer space and
+ *     more data available
+ *
+ * Options:
+ *   replace_chars_stdout, replace_chars_stderr - remove non-printable
+ *     characters from stdout/stderr
+ */
+int handle_remote_data(
+    libvchan_t *data_vchan, int stdin_fd, int *status,
+    struct buffer *stdin_buf, int data_protocol_version,
+    bool replace_chars_stdout, bool replace_chars_stderr);
+
+/*
+ * Handle data from the specified FD (cannot be -1) and send it over vchan
+ * with a given message type (MSG_DATA_STDIN/STDOUT/STDERR).
+ *
+ * Return codes:
+ *   REMOTE_ERROR - vchan error occured
+ *   REMOTE_EOF - EOF received, do not access this FD again
+ *   REMOTE_OK - some data processed, call it again when buffer space and
+ *     more data availabla
+ */
+int handle_input(
+    libvchan_t *vchan, int fd, int msg_type,
+    int data_protocol_version);
+
+int send_exit_code(libvchan_t *vchan, int status);
+
+/* Set of options for process_io(). */
+struct process_io_request {
+    libvchan_t *vchan;
+    struct buffer *stdin_buf;
+
+    // stderr_fd can be -1
+    int stdin_fd, stdout_fd, stderr_fd;
+    // 0 if no child process
+    pid_t local_pid;
+
+    /*
+      is_service true (this is a service):
+        - send local data as MSG_DATA_STDOUT
+        - send exit code
+        - wait just for local end
+        - return local exit code
+
+      is_service false (this is a client):
+        - send local data as MSG_DATA_STDIN
+        - don't send exit code
+          (TODO: should it be this way?)
+        - if there's no process, wait for remote end
+        - return local or remote exit code, whichever finished first
+          (TODO: should it be this way?)
```

Comment:
It should not depend on their exit order. As for which exit code return - the reasoning for returning local one (if local process present) is that local process may have better chance to interpret remote data and handle various expected failures gracefully (but it have no way to receive remote process exit code directly - can only use in-band data). On the other hand, this sometimes returning remote exit code and sometimes not is pretty confusing, so it may worth unifying it (and always return remote exit code).
I don't think this change would break anything, as local process is used rarely. But if we'd go with the change, we need to verify this.